### ISSUE TITLE
[PHP] Add CURLOPT_PROXY options

### ISSUE THREAD

**Issue by maik-kulbe** [swagger-api/swagger-codegen#4071]
##### Description;

Using the generated code behind a corporate firewall without direct access to the API Endpoint server breaks the generated client code because 
##### Swagger-codegen version

2.2.0
##### Command line used for generation

Tested with generated PHP code
##### Steps to reproduce

Run generated code on a machine without route to the API Endpoint host, but with an HTTP proxy enabled(e.g. on a remote server)
##### Related issues
- [PHP] add CURLOPT_TCP_NODELAY option #3962
  - Asks for another curl option
##### Suggest a Fix

Generated code should allow for HTTP or SOCKS proxy options to be set. This includes the following:
- Proxy schema (http/https) + Proxy host -> `CURLOPT_PROXY`
- Proxy port -> `CURLOPT_PROXYPORT`
- Proxy type(HTTP/SOCKS5) -> `CURLOPT_PROXYTYPE`, e.g. set to `CURLPROXY_HTTP`
- Proxy user [optional] + Proxy password [optional] -> `CURLOPT_PROXYUSERPWD`

``` php
// Something similar to this(added error handling, options not being set, ..) would need 
// to be added to modules/swagger-codegen/src/main/resources/php/ApiClient.mustache
$options = array();
$options[CURLOPT_PROXY] = $this->config->getProxyHost();
$options[CURLOPT_PROXYPORT] = $this->config->getProxyPort();
$options[CURLOPT_PROXYTYPE] = $this->config->getProxyType();
$options[CURLOPT_PROXYUSERPWD] = $this->config->getProxyUser() . ":" . 
                                 $this->config->getProxyPassword();

curlopt_setopt_array($curl, $options);
```

**Comment by wing328** [comment]
@maik-kulbe your suggestion sounds reasonable to me. May I know if you've time to contribute the enhancement?

**Comment by wing328** [comment]
cc @arnested

**Comment by maik-kulbe** [comment]
@wing328 I'll try to, but I don't have a very deep understanding of this repos structure, so if I PR a change later it might need some work.

**Comment by wing328** [comment]
@maik-kulbe thanks. A good starting point is to review the merged PRs for PHP API client: https://github.com/swagger-api/swagger-codegen/pulls?q=is%3Apr+label%3A%22Client%3A+PHP%22+is%3Aclosed

**Comment by maik-kulbe** [comment]
Sorry this took me so long, but didn't find any time to tackle this until now.

I've added the options after patching the generated code and working my way backwards to the template files, but I'm strugglng with the test setup you got for this codebase. Is there any extensive guide on how to set everything up? I'm not familiar with Maven or Java and at the moment I can't even get the script under /bin running...

**Comment by wing328** [comment]
@maik-kulbe No worry. Please submit the PR and I'll update the PHP Petstore sample for you.

**Comment by wing328** [comment]
Closed via #4596. Thanks for the PR from @maik-kulbe
